### PR TITLE
Fix `BFields` of generated immutable `BRecords` not being marked as `SymbolFlags.READONLY`

### PR DESCRIPTION
**Pull request by Thushara-Piyasekara**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/42520

## Approach
After a discussion with @heshanpadmasiri, it was apparent the `InherentTypeViolation` error for `BFields` of readonly `BRecords` were not getting triggered. This was due to the mutable `BRecord` generation not inheriting the readonly flag to the generated `BFields`.

After refactoring the logic to make the `BFields` also inherit the readonly flag, both cases mentioned in https://github.com/ballerina-platform/ballerina-lang/issues/42520 triggered the same error.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package org.ballerinalang.test.types.readonly;
```
Comment (nipunayf) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:1]: Missing the header

Comment (Thushara-Piyasekara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:19]: Fixed in 6e9b1d9980383e38660e4e1193ee3a3b647a8928. It's an obvious mistake in my part. Will make sure to double check in the future.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.ballerinalang.test.CompileResult;
+import org.ballerinalang.test.exceptions.BLangTestException;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Tests for the readonly violation error consistency between BTuple type and BRecord type.
+ *
+ * @since 2201.9.0
```
Comment (warunalakshitha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:32]: This needs to be updated

Comment (Thushara-Piyasekara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java]: Fixed in 1dbfc97b0b6380de20e2db72032beaf4af057b5e.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java [hunk 3]
```diff
@@ -169,9 +169,10 @@ public void testCreateRawDetails() {
     }
 
     @Test(expectedExceptions = BLangTestException.class,
-          expectedExceptionsMessageRegExp = ".*Invalid update of record field: modification not allowed on readonly " +
-                  "value.*")
+            expectedExceptionsMessageRegExp = "error: \\{ballerina/lang.map}InherentTypeViolation \\{\"message\":" +
+                    "\"cannot update 'readonly' field 'name' in record of type 'Details & readonly'\".*")
     public void testCreateDetails() {
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java:175]: Unnecessary change.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Tests for the readonly violation error consistency between BTuple type and BRecord type.
+ *
+ * @since 2201.10.0
+ */
+public class TupleVsArrayReadonlyErrorConsistencyTest {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:34]: Do we need to mention consistency here? I'd just include readonly field update.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        resultWithoutTupleUpdateMethod = BCompileUtil.compile("test-src/types/readonly/" +
+                "test_tuple_vs_array_readonly_violation_consistency_without_tuple_Update_method.bal");
+    }
+
+    @Test(expectedExceptions = BLangTestException.class,
+            expectedExceptionsMessageRegExp = "error: \\{ballerina/lang.map}InherentTypeViolation \\{\"message\":" +
+                    "\"cannot update 'readonly' field 'name' in record of type '\\(Employee & readonly\\)'\".*")
+    public void testWithTupleUpdateMethod() {
+        BRunUtil.invoke(resultWithTupleUpdateMethod, "testFrozenAnyArrayElementUpdate");
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:51]: We can trap the panic and do the assertion in Ballerina itself, at least for new tests.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java (hunk 4 above)
Comment (Thushara-Piyasekara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:34]: The "consistency" keyword was added because the issue being fixed by this PR https://github.com/ballerina-platform/ballerina-lang/issues/42520 is related to an inconsistency between the errors produced during runtime. 

Are you suggesting to modify the test case to match the actual bug? or is it better to use the existing real world test case?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java (hunk 5 above)
Comment (Thushara-Piyasekara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:51]: Shall we use this as the new test and assert the exit code from java side?

```ballerina
public function main() returns error? {
    error? actualError = trap testFrozenAnyArrayElementUpdate();
    if actualError is error {
        test:assertEquals(actualError.message(), "{ballerina/lang.map}InherentTypeViolation");
    }
}
```

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:51]: The test has to fail if `actualError` is not an error too.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java (hunk 4 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:34]: I would just have them be part of readonly field tests. If this was the behaviour to start with, we wouldn't have had a consistency issue.

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/readonly/TupleVsArrayReadonlyErrorConsistencyTest.java:34]: Will add them to the place where readonly field tests were defined

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java [hunk 6]
```diff
@@ -38,6 +38,8 @@ public class MappingConstructorExprTest {
     private CompileResult inferRecordResult;
     private CompileResult spreadOpFieldResult;
     private CompileResult readOnlyFieldResult;
+    private CompileResult resultWithTupleUpdateMethod;
+    private CompileResult resultWithoutTupleUpdateMethod;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java:42]: Can't the tests be merged to a single file, with different functions?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java [hunk 7]
```diff
@@ -47,6 +49,10 @@ public void setup() {
         varNameFieldResult = BCompileUtil.compile("test-src/expressions/mappingconstructor/var_name_field.bal");
         spreadOpFieldResult = BCompileUtil.compile("test-src/expressions/mappingconstructor/spread_op_field.bal");
         readOnlyFieldResult = BCompileUtil.compile("test-src/expressions/mappingconstructor/readonly_field.bal");
+        resultWithTupleUpdateMethod =
+                BCompileUtil.compile("test-src/types/readonly/test_tuple_vs_array_readonly_violation_consistency.bal");
+        resultWithoutTupleUpdateMethod = BCompileUtil.compile("test-src/types/readonly/" +
+                "test_tuple_vs_array_readonly_violation_consistency_without_tuple_Update_method.bal");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java:55]: consistency doesn't make sense for the file name.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java (hunk 6 above)
Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java:42]: Make them local variables anyway. Add comments to say what's being tested.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java [hunk 8]
```diff
@@ -169,8 +169,8 @@ public void testCreateRawDetails() {
     }
 
     @Test(expectedExceptions = BLangTestException.class,
-          expectedExceptionsMessageRegExp = ".*Invalid update of record field: modification not allowed on readonly " +
-                  "value.*")
+            expectedExceptionsMessageRegExp = "error: \\{ballerina/lang.map}InherentTypeViolation \\{\"message\":" +
+                    "\"cannot update 'readonly' field 'name' in record of type 'Details & readonly'\".*")
     public void testCreateDetails() {
         BRunUtil.invoke(result, "testCreateDetails");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java:175]: If the entire record is immutable, the former error is better?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/readonly/test_tuple_vs_array_readonly_violation_consistency.bal [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+type Employee record {|
+    int id;
+    string name;
+|};
+
+function testFrozenAnyArrayElementUpdate() returns error? {
+    error? actualError = trap frozenAnyArrayElementUpdate();
+    test:assertTrue(actualError is error);
+    test:assertEquals((<error>actualError).message(), "{ballerina/lang.map}InherentTypeViolation");
+    test:assertEquals((<error>actualError).detail()["message"],
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/readonly/test_tuple_vs_array_readonly_violation_consistency.bal:28]: Extract `<error>actualError` out to a variable.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java (hunk 6 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/mappingconstructor/MappingConstructorExprTest.java:42]: Since this is a error that happen when code generation, It is needed to be in two files

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java (hunk 8 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/javainterop/basic/StaticMethodTest.java:175]: The current error message is the one we normally get, when we updating a immuatable value

**Review by gimantha — APPROVED** [review]
